packages/core-contracts/contracts/utils/ERC20.sol

```diff
@@ -0,0 +1,101 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "../interfaces/IERC20.sol";
+
+abstract contract ERC20 is IERC20 {
+    string public override name;
+
+    string public override symbol;
+
+    uint8 public immutable override decimals;
+
+    mapping(address => uint256) public override balanceOf;
+
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    uint256 public override totalSupply;
+
+    constructor(
+        string memory tokenName,
+        string memory tokenSymbol,
+        uint8 tokenDecimals
+    ) {
+        name = tokenName;
+        symbol = tokenSymbol;
+        decimals = tokenDecimals;
+    }
+
+    function approve(address spender, uint256 amount) public override returns (bool) {
+        allowance[msg.sender][spender] = amount;
+
+        emit Approval(msg.sender, spender, amount);
+
+        return true;
+    }
+
+    function transfer(address to, uint256 amount) public override returns (bool) {
+        _transfer(msg.sender, to, amount);
+
+        return true;
+    }
+
+    function transferFrom(
+        address from,
+        address to,
+        uint amount
+    ) external override returns (bool) {
+        uint256 currentAllowance = allowance[from][msg.sender];
+        // it reverts if the sender does not have sufficient allowance
+        require(currentAllowance >= amount, "Amount exceeds allowance");
+        unchecked {
+            allowance[from][msg.sender] -= amount;
```

Comment:
Shouldn't any changes in approval emit Approval events?
(not sure what the spec has to say about this)